slackapi/deno-slack-api
# Add trigger response types, accept workflows as a generic

##  Summary

This PR introduces two features and a ton of tests
1. Trigger methods now receive workflows as an optional generic. This provides typing for the `workflow` field and it powers the `inputs` field. It's used by both CLI files and at runtime.
2. Triggers now have response typing. This is only used at runtime.

```ts
// manifest.ts
export const MixedWorkflow = DefineWorkflow({
  callback_id: "mixed",
  title: "Example",
  input_parameters: {
    properties: {
      optional: {
        type: "string",
      },
      required: {
        type: "string",
      },
    },
    required: ["required"],
  },
});

// reverse.ts
const resp = await client.workflows.triggers.create<typeof MixedWorkflow.definition>({
  inputs: {} // <- This should now be typed
  workflow: // <- This should now be typed 
})

if (resp.ok) {
  resp.trigger; // <- This should now be typed
}
```

#### CLI
```ts
const trigger: Trigger<typeof ReverseWorkflow.definition> = {
  name: "Test Trigger",
  type: "shortcut",
  workflow: "#/workflows/reverse_wf",
};
```
#### Runtime
```ts
await client.workflows.triggers.create<typeof ReverseWorkflow.definition> = {
  name: "Test Trigger",
  type: "shortcut",
  workflow: "#/workflows/reverse_wf",
}
```

#### Testing

Since these are just type changes testing should mostly be done against the experience to write triggers, but it's always good to make sure things are working e2e
* Test with and without generics
* Test workflows that require inputs, workflows that don't have any inputs, workflows that have optional inputs, etc.
* Test to make sure that `workflow` param is always right
* Verify the response is what you'd expect for the different trigger types
* Try to find any case where we're not providing flexibility where we should

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/deno-slack-api/blob/main/.github/CONTRIBUTING.md) and have done my best effort to follow them.
